packages/falcon-server/src/schemaDirectives/GraphQLCacheDirective.js

```diff
@@ -0,0 +1,269 @@
+    const { name: returnTypeName } = this.getRootType(info.returnType);
+    const tags = [returnTypeName];
+
+    // Checking if Type is "self-cacheable"
+    tags.push(...this.getTagsForField(resolverResult, info.returnType));
+
+    idPath.forEach(idPathEntry => {
+      tags.push(...this.extractTagsForIdPath(idPathEntry, resolverResult, info, parent));
+    });
+
+    return {
+      value: resolverResult,
+      options: {
+        ...((result && result.options) || {}),
+        tags
+      }
+    };
+  }
+
+  /**
+   * Extract cache tags for the provided ID path and return a list of tags
+   * @param {string} idPath ID operation path string (like "$parent.items" or "items")
+   * @param {object} result Resolver result
+   * @param {object} info GraphQL info object
+   * @param {object} parent GraphQL parent object
+   * @return {string[]} List of tags
+   */
+  extractTagsForIdPath(idPath, result, info, parent) {
+    const [rootPath, ...pathParts] = idPath.split(PATH_SEPARATOR);
+    const valueToCheck = rootPath === PARENT_KEYWORD ? parent : result;
+    const typeToCheck = this.getRootType(rootPath === PARENT_KEYWORD ? info.parentType : info.returnType);
+    if (rootPath !== PARENT_KEYWORD) {
+      // Put first path section back to "pathParts" for non-parent entries
+      pathParts.unshift(rootPath);
+    }
+
+    return this.getTagsForField(valueToCheck, typeToCheck, pathParts);
+  }
+
+  /**
+   * Get a list of tags from the provided value for the specified typeField
+   * @param {object} value Value to checks
+   * @param {object} typeField GraphQL Type Object
+   * @param {string[]} [pathParts=[]] Path parts
+   * @return {string[]} List of tags
+   */
+  getTagsForField(value, typeField, pathParts = []) {
+    if (!pathParts.length) {
+      const { name: typeName } = this.getRootType(typeField);
+      return this.generateCacheTags(typeName, this.extractFieldValue(value, this.findTagIdFieldName(typeField)));
+    }
+
+    const [currentPath, ...restIdPath] = pathParts;
+    const { _fields: fields } = typeField;
+    let { name: typeName } = typeField;
+    let fieldValue = this.extractFieldValue(value, currentPath);
+
+    // Keep looking for nested ID path until it reaches the end node
+    if (currentPath && restIdPath.length) {
+      return this.getTagsForField(fieldValue, this.getRootType(fields[currentPath].type), restIdPath);
+    }
+    if (Array.isArray(fieldValue)) {
+      const currentType = this.getRootType(fields[currentPath].type);
+      typeName = currentType.name;
+      const currentCacheIdFieldName = this.findTagIdFieldName(currentType);
+      fieldValue = this.extractFieldValue(fieldValue, currentCacheIdFieldName);
+    }
+
+    return [typeName, ...this.generateCacheTags(typeName, fieldValue)];
+  }
+
+  /**
+   * Find a field name with TAG_ID_FIELD_TYPE type
+   * @param {object} objectType GQL Object Type
+   * @return {string|undefined} Name of the field with
+   */
+  findTagIdFieldName(objectType) {
```

Comment:
what happens when `objectType` will have more than one field which type is `TAG_ID_FIELD_TYPE` (composite identifier)?  looks like the only first one fieldName will be returned which will be wrong in that case I think.

should not be `typeObject` or `gqlType`?
```suggestion
  findTagIdFieldName(typeObject) {
```

Reply:
updated